chore: split `awkward` into two packages

agoose77 opened the pull request:
This PR closes #503, fixes #1599

## High Level
This PR makes the boundary between the C++ portion of Awkward Array and the Python-only portion much sharper. This has benefits for the ease of caching in the CI, but also helps us make quicker, more lightweight releases; and perform editable installs.

Structurally, data that is shared by `awkward` and `awkward-cpp` is placed at the Git root. This makes it easy to see the relations between the packages. Dev scripts prepare both packages for sdist generation using `nox -s prepare`. We do not include the dev scripts or kernel-specification into the sdists, therefore they should not be used during wheel building. Only the generated files (tests, headers) are included in the sdists and hence wheels.

## Tasks
- [x] Split awkward and awkward-core packages
- [x] Use `awkward-cpp` name
- [x] Use strict equality version constraint
- [x] Update CI 
   - [x] Only run build for changes to `awkward-core`
   - [x] Use cached awkward-core wheels using hash of awkward-core
- [x] Remove `localbuild.py`
- [x] Move `./awkward-cpp/cpp-headers` to `./header-only`, included as implementation details into two package sdists (and pure-python wheel too).
- [x] Move `./src/awkward/cpp-headers/rdataframe/` to `./src/awkward/_connect/rdataframe/include` (approx)
- [x] Clean-up MANIFEST.in
- [x] Update LayoutBuilder user-guide to reflect best practices for locating headers
- [x] Update `contributing.md`
- [x] Move doxygen docs to `awkward-cpp/docs`, and include in sdist
- [x] Rename `docs-sphinx` to `docs`
- [x] Add helpful sessions to `nox` for code generation
- [x] Check for out-of-sync awkward and awkward-cpp versions/requirements ON RELEASE
- [x] Check for modifications to released awkward-cpp versions ON RELEASE
- [x] Move version source to `pyproject.toml`
- [x] Move contributing info to details
- [x] Tie public C++ headers to Python package version
- [x] Ensure that sdists can be built into functional wheels
- [ ] Embed C++ header artefact as a GH release asset
- [x] Optimise kernel test generation
- [ ] Use https://github.com/softprops/action-gh-release to create releases _from_ workflows

<!-- docs-preview-start -->
----
:books: The documentation for this PR will be available at <https://awkward-array.readthedocs.io/en/agoose77-chore/split-awkward-package/> once Read the Docs has finished building :hammer:
<!-- docs-preview-end -->

agoose77 committed:
chore: move submodules to libawkward

agoose77 committed:
chore: remove include to libawkward

agoose77 committed:
chore: remove kernel specs to libawkward

agoose77 committed:
chore: move flake8 config to .flake8

agoose77 committed:
chore: move tests-cpp to libawkward

agoose77 committed:
chore: move src/... to libawkward/src

agoose77 committed:
chore: only compile tests in libawkward

agoose77 committed:
chore: move awkward cpp-headers to libawkward

agoose77 committed:
chore: move setup files

agoose77 committed:
chore: add pyproject.toml for awkward package

agoose77 committed:
chore: add VERSION_INFO for libawkward

agoose77 committed:
chore: reflect split cpp-headers in CMakeLists.txt

agoose77 committed:
chore: hard-code version in awkward

agoose77 committed:
chore: add pyproject.toml for cibuildwheel

agoose77 committed:
chore: add LICENSE

agoose77 committed:
chore: use `libawkward` name

agoose77 committed:
chore: generate kernels in correct location

agoose77 committed:
chore: don't output libraries in install dir

agoose77 committed:
chore: restructure library loading

agoose77 committed:
chore: cleanup setup.py

agoose77 committed:
chore: read dependencies from txt files

agoose77 committed:
chore: use awkward_core dependency

agoose77 committed:
chore: rename libawkward to awkward-core

agoose77 committed:
chore: use `awkward-cpp` name

Use single-digit version, fix kernel signature generation

agoose77 committed:
fix: refer to `awkward_cpp.libawkward`

agoose77 committed:
Merge branch 'main' into agoose77/chore/split-awkward-package

agoose77 committed:
fix: add pyyaml and cmake to pyproject.toml

agoose77 committed:
fix: correct path to `__init__`

agoose77 committed:
fix: correct imports for new location of `_ext`

jpivarski committed:
Baseline Python package version should be rc3 and include a dev number.

jpivarski committed:
Resolve merge conflicts with main in all C++ files.

jpivarski committed:
Merge branch 'main' into agoose77/chore/split-awkward-package

jpivarski committed:
Include the C++ build step in GitHub Actions.

agoose77 committed:
wip: move manifest

agoose77 committed:
chore: remove config.py

These headers should no-longer be resolved via the `awkward` package

jpivarski committed:
Fix kernel test execution for new locations.

jpivarski committed:
Merge branch 'agoose77/chore/split-awkward-package' of https://github.com/scikit-hep/awkward into agoose77/chore/split-awkward-package

pre-commit-ci[bot] committed:
style: pre-commit fixes

agoose77 committed:
refactor: move header-only headers to `src/awkward/_connect`

agoose77 committed:
ci: generate required files before building package

agoose77 committed:
ci: install all requirements

agoose77 committed:
ci: be smarter, requirements before usage -.-

agoose77 committed:
chore: fix copy-cpp-headers for python<3.8

agoose77 committed:
ci: run preparatory scripts for ROOT build

agoose77 committed:
chore: remove generation code from setup.py

agoose77 committed:
chore: fix moved VERSION_INFO

agoose77 committed:
chore: remove VERSION_INFO

agoose77 committed:
chore: restore VERSION_INFO

agoose77 committed:
chore: let version be a single digit

agoose77 committed:
chore: use scikit-build-core to simplify windows build

agoose77 committed:
chore: remove version info

agoose77 committed:
Merge branch 'main' of https://github.com/scikit-hep/awkward into agoose77/chore/split-awkward-package

pre-commit-ci[bot] committed:
style: pre-commit fixes

agoose77 committed:
fix: pin scikit-build-core, require ninja

agoose77 committed:
fix: specify package lower bound

agoose77 committed:
fix: don't try to drop const

agoose77 committed:
refactor: remove unused import

agoose77 committed:
chore: cleanup sdist contents

agoose77 committed:
chore: include tests-cuda-kernels

pre-commit-ci[bot] committed:
style: pre-commit fixes

agoose77 committed:
chore: fix pyproject.toml

agoose77 committed:
chore: don't define test and dev dependencies

agoose77 committed:
chore: define cibuildwheel configuration in separate file

agoose77 committed:
Merge remote-tracking branch 'refs/remotes/origin/agoose77/chore/split-awkward-package' into agoose77/chore/split-awkward-package

pre-commit-ci[bot] committed:
style: pre-commit fixes

agoose77 committed:
chore: fix pylint nox session

agoose77 committed:
Merge remote-tracking branch 'refs/remotes/origin/agoose77/chore/split-awkward-package' into agoose77/chore/split-awkward-package

agoose77 committed:
chore: remove pybind11 submodule

agoose77 committed:
wip: use patched scikit-build-core for wasm build

agoose77 committed:
ci: wip docs building

agoose77 committed:
docs: fix path to header-only includes

agoose77 committed:
ci: invoke preparatory steps in docs building

agoose77 committed:
ci: install pyyaml

agoose77 committed:
chore: update paths in debug statements

agoose77 committed:
chore: bump scikit-build-core commit

agoose77 committed:
ci: fix path in docs ci

agoose77 committed:
ci: try caching wasm-wheel build

agoose77 committed:
ci: fix docs copy operation

agoose77 committed:
chore: drop use of ccache in favour of wheel caching

agoose77 committed:
ci: wip update to use cached wheels rather than ccache

agoose77 committed:
ci: run preparatory stage before wheel cache

agoose77 committed:
ci: fix cache key job dependence

agoose77 committed:
ci: don't run build with pipx

agoose77 committed:
ci: don't run build with pipx

agoose77 committed:
ci: pass `environment-name` for `environment-file: false`

agoose77 committed:
ci: use backtick line continuation on windows

agoose77 committed:
chore: remove autograd

agoose77 committed:
ci: fix paths for windows (?)

agoose77 committed:
ci: specify conda-forge for ROOT install

agoose77 committed:
ci: remove ./ prefix

agoose77 committed:
chore: allow more packages on 3.11

agoose77 committed:
ci: bump NumPy, JAX minimum version

agoose77 committed:
ci: install build from PyPI (up-to-date)

agoose77 committed:
ci: try ignoring dotfiles in hash computation

agoose77 committed:
ci: fix `channels` specifier

agoose77 committed:
ci: enable predictable hash

agoose77 committed:
ci: remove special-case for dotfiles

agoose77 committed:
ci: use more recent epoch time :)

agoose77 committed:
ci: remove unused variable

agoose77 committed:
ci: switch to `python` on windows(?)

agoose77 committed:
chore: fix path in generate-kerneldocs.py

agoose77 committed:
ci: debug windows build

agoose77 committed:
ci: debug windows build (2)

agoose77 committed:
ci: try fixing install

agoose77 committed:
ci: try explicit wildcard resolution

agoose77 committed:
chore: simplify requirements

agoose77 committed:
chore: simplify dev scripts and workflow

agoose77 committed:
ci: use only `requirements-test.txt`

agoose77 committed:
feat: add `ak.builder` module to re-export `ArrayBuilder`

agoose77 committed:
chore: add exports to __init__

agoose77 committed:
chore: move `ak._ext.ArrayBuilder.rst`

agoose77 committed:
docs: make C++ more self-contained

agoose77 committed:
docs: don't generate kernel documentation in Sphinx conf.py

agoose77 committed:
docs: move docs-sphinx to `docs`

agoose77 committed:
chore: improve docs task in nox

agoose77 committed:
ci: update CI to build docs separately

agoose77 committed:
chore: reorder noxfile for docs

agoose77 committed:
chore: remove PyPI README

agoose77 committed:
chore: add gitignore, remove MANIFEST.in

agoose77 committed:
ci: remove check-manifest

agoose77 committed:
chore: update contributing information in README

agoose77 committed:
chore: install pytest-xdist

agoose77 committed:
docs: work on CONTRIBUTING

agoose77 committed:
Merge branch 'main' into agoose77/chore/split-awkward-package

agoose77 committed:
docs: mention release artifacts on GitHub for headers

agoose77 committed:
docs: fix paths to awkward-cpp tests

agoose77 committed:
ci: fix missing dependency

agoose77 committed:
ci: fix generation call

agoose77 committed:
ci: fix RTD

agoose77 committed:
chore: use scikit-build upstream for scikit-build-core

agoose77 committed:
docs: fix kernel doc generation call

agoose77 committed:
ci: download new artefact names in RTD

agoose77 committed:
chore: make generated docs directory if it doesn't exist

agoose77 committed:
ci: include arch in cache key

agoose77 committed:
ci: first pass at tests workflow

agoose77 committed:
ci: fix cibuildwheel test invocation

agoose77 committed:
ci: build and publish pure-python wheel

agoose77 committed:
ci: specify cibuildwheel `package-dir` option

agoose77 committed:
ci: fix docs slug computation & check CURL exit status

agoose77 committed:
docs: require awkward>2.0.0

agoose77 committed:
ci: wip total release work

agoose77 committed:
docs: require >=2.0.0rc0 for awkward

agoose77 committed:
ci: wip for separate deployment workflows

agoose77 committed:
chore: fix RTD regex for non-WASM wheel

agoose77 committed:
chore: run doxygen on RTD

agoose77 committed:
docs: improve CONTRIBUTING.md

agoose77 committed:
chore: add dev script to check awkward-cpp is unmodified

agoose77 committed:
ci: update ignored paths

agoose77 committed:
ci: add version-check workflow

agoose77 committed:
ci: fix version-check workflow

agoose77 committed:
ci: fix version-check workflow (2)

agoose77 committed:
ci: fix version-check workflow (2)

agoose77 committed:
ci: add missing toml dependency

agoose77 committed:
ci: remove yaml dependency

agoose77 committed:
chore: add dev-script to check package requirements match

agoose77 committed:
ci: also check packaging requirements are compatible

agoose77 committed:
chore: add check_version session to nox

agoose77 committed:
ci: use nox where possible

agoose77 committed:
docs: update contributing

agoose77 committed:
ci: use nox to run kernel diagnostics

agoose77 committed:
chore: use same `requirements-dev.txt` for dev sessions in nox

agoose77 committed:
chore: use same `requirements-dev.txt` for dev sessions in nox

agoose77 committed:
Merge remote-tracking branch 'origin/agoose77/chore/split-awkward-package' into agoose77/chore/split-awkward-package

agoose77 committed:
ci: use `pipx run build`

agoose77 committed:
chore: fix nox arg parsing

agoose77 committed:
chore: use argparse in nox

agoose77 committed:
ci: fix Linux workflow

agoose77 committed:
ci: revert back to `python -m build`

henryiii commented on awkward-cpp/pyproject.toml:
```diff
@@ -0,0 +1,58 @@
+	"scikit-build-core @ git+https://github.com/scikit-build/scikit-build-core@887757e85d728bad97fb883c0cf7c132298e32f6",
+    "pybind11",
+    "ninja==1.10.2.4; platform_system!='Windows'",
+    "cmake"
```
```suggestion
```

You should never specify cmake and ninja when using scikit-build-core, it will automatically add those only if the system versions are inadequate. This way you can build on systems that do not provide pre-built wheels!

agoose77 replied:
That's great news, I'll remove them.

henryiii commented on awkward-cpp/CMakeLists.txt:
```diff
@@ -111,31 +111,21 @@ if(BUILD_TESTING)
+target_link_libraries(_ext PRIVATE awkward-static)
+
+# Install pure-python files
+file(GLOB_RECURSE PYTHON_SOURCES "src/${SKBUILD_PROJECT_NAME}/*.py")
```
If this is named correctly (which it looks like it is), the next version of scikit-build-core (feel free to bump the sha, already in master for matching named packages) will copy the files over for you. So `src/<package>` or `<package>` gets copied to `<package>`, similar to most other tools.

I'll probably make a next release today or tomorrow.

agoose77 replied:
That's really useful. I'll bump the SHA for now.

agoose77 committed:
ci: fix Windows workflow

agoose77 committed:
Update awkward-cpp/pyproject.toml

Co-authored-by: Henry Schreiner <henry.fredrick.schreiner@cern.ch>

agoose77 committed:
chore: use scikit-build-core's copying logic

agoose77 committed:
ci: fix windows path

henryiii commented on awkward-cpp/CMakeLists.txt:
```diff
@@ -111,31 +111,21 @@ if(BUILD_TESTING)
+  RUNTIME DESTINATION "${SKBUILD_PROJECT_NAME}/lib"
+  ARCHIVE DESTINATION "${SKBUILD_PROJECT_NAME}/lib")
+
+install(FILES ${PYTHON_SOURCES} DESTINATION ${SKBUILD_PROJECT_NAME})
```
Don't forget this one :)

agoose77 replied:
Oh, good catch - I didn't read the line closely enough.

agoose77 committed:
chore: remove install command!

agoose77 committed:
Revert "chore: remove install command!"

This reverts commit 8dd90983a3a1a0a2878c5db73962750fef085dcb.

agoose77 committed:
Revert "chore: use scikit-build-core's copying logic"

This reverts commit 779c61cdb3a45b61a64bfd45449dcb4396ebdf82.

agoose77 committed:
chore: move C++ tests to self-contained path

agoose77 committed:
ci: set SOURCE_DATE_EPOCH for C++ deployment to git changed date

agoose77 committed:
ci: don't set SOURCE_DATE_EPOCH for pure-Python package

agoose77 committed:
ci: set SOURCE_DATE_EPOCH for pure-Python package

agoose77 committed:
ci: don't run tests on cpp if cache hits

agoose77 committed:
chore: fix RTD

agoose77 committed:
chore: fix cache key for windows

agoose77 committed:
fix: bundle artefacts

agoose77 committed:
fix: reproducible sdist

agoose77 committed:
fix: make tests reproducible too

jpivarski reviewed (commented):
I followed the most recent instructions, https://github.com/scikit-hep/awkward/pull/1778#issuecomment-1305825665, in an environment that has ROOT (for the ROOT tests) and no pre-existing `awkward` or `awkward-cpp`.

(Even though I have git `submodule.recurse=true` now, I still did `git clone --recursive https://github.com/scikit-hep/awkward.git -b agoose77/chore/split-awkward-package` anyway. Why not?)

These are the steps:

```bash
python dev/copy-cpp-headers.py
python dev/generate-kernel-signatures.py
pip install -vv ./awkward-cpp
pip install -e .
pytest -n 16 tests
```

The `awkward-cpp` compilation was slower than usual, 2m30s, because it was not parallelized. It lost the parallel compilation argument.

One test failed:

```
tmp_path = PosixPath('/tmp/pytest-of-jpivarski/pytest-0/popen-gw12/test_data_frame_vec_of_vec_of_0')

    def test_data_frame_vec_of_vec_of_real(tmp_path):
        filename = os.path.join(tmp_path, "test-listarray.root")
    
        ak_array_in = ak.Array([[[1.1], [2.2]], [[3.3], [4.4, 5.5]]])
    
        data_frame = ak.to_rdataframe({"x": ak_array_in})
    
        assert data_frame.GetColumnType("x").startswith("awkward::ListArray_")
    
        ak_array_out = ak.from_rdataframe(
            data_frame,
            columns=("x",),
        )
        assert ak_array_in.to_list() == ak_array_out["x"].to_list()
    
        with pytest.raises(SystemError):
>           data_frame.Snapshot("ListArray", filename, ("x",))
E           Failed: DID NOT RAISE <class 'SystemError'>

ak_array_in = <Array [[[1.1], [2.2]], [[3.3], ...]] type='2 * var * var * float64'>
ak_array_out = <Array [{x: [[1.1], [2.2]]}, {...}] type='2 * {x: var * var * float64}'>
data_frame = <cppyy.gbl.ROOT.RDF.RInterface<ROOT::Detail::RDF::RLoopManager,void> object at 0x55976e347090>
filename   = '/tmp/pytest-of-jpivarski/pytest-0/popen-gw12/test_data_frame_vec_of_vec_of_0/test-listarray.root'
tmp_path   = PosixPath('/tmp/pytest-of-jpivarski/pytest-0/popen-gw12/test_data_frame_vec_of_vec_of_0')

tests/test_1781-rdataframe-snapshot.py:53: Failed
```

@ianna, is this something wrong with the ROOT setup? I seem to remember it acting like this at some point, but I can't remember when or why.

It is _not_ the error I reported before, in which ROOT failed with a segfault.

Uh oh—I ran all of the tests a few more times (parallel and serial) and the above error did not recur. It _is_ the first time I'm using ROOT after a new installation, so it is possible that this is due to something that gets configured once. I guess it's beyond the scope of this PR, but there's something a bit flaky about this particular part: raising SystemError when attempting to `Snapshot` an unsnapshotable ArrayView in RDataFrame.

ianna commented on CONTRIBUTING.md:
````diff
@@ -52,31 +52,66 @@ It is unnecessary to manually edit (rebase) your commit history. If, however, yo
+Then the installation can be performed without build isolation:
+```bash
+python -m pip install --no-build-isolation ./awkward-cpp
+```
````
```diff
 % python -m pip install --no-build-isolation ./awkward-cpp
Processing ./awkward-cpp
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [19 lines of output]
      Traceback (most recent call last):
        File "/Users/yana/opt/anaconda3/envs/my-ppp-talk/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 156, in prepare_metadata_for_build_wheel
          hook = backend.prepare_metadata_for_build_wheel
      AttributeError: module 'scikit_build_core.build' has no attribute 'prepare_metadata_for_build_wheel'
      
      During handling of the above exception, another exception occurred:
      
      Traceback (most recent call last):
        File "/Users/yana/opt/anaconda3/envs/my-ppp-talk/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 363, in <module>
          main()
        File "/Users/yana/opt/anaconda3/envs/my-ppp-talk/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 345, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "/Users/yana/opt/anaconda3/envs/my-ppp-talk/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 160, in prepare_metadata_for_build_wheel
          whl_basename = backend.build_wheel(metadata_directory, config_settings)
        File "/Users/yana/opt/anaconda3/envs/my-ppp-talk/lib/python3.10/site-packages/scikit_build_core/build.py", line 24, in build_wheel
          from .pyproject.wheel import build_wheel as skbuild_build_wheel
        File "/Users/yana/opt/anaconda3/envs/my-ppp-talk/lib/python3.10/site-packages/scikit_build_core/pyproject/wheel.py", line 10, in <module>
          from pyproject_metadata import StandardMetadata
      ModuleNotFoundError: No module named 'pyproject_metadata'
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.

```

agoose77 replied:
@ianna that's ... odd. Could you run `pip list -vvv`?

agoose77 replied:
Wait, no that's probably because the latest version of `scikit-build-core` isn't on PyPI yet. You'd need to install the same version as specified in `awkward-cpp`'s `pyproject.toml` until the release is made!

henryiii replied:
It's because `pyproject_metadata`, `distlib`, and `pathspec` are also dependencies of scikit-build-core (but only when building wheels and/or sdists).

henryiii replied:
(As well needing to install the git version until I release, that too)

agoose77 replied:
@henryiii right, you add them as additional requires - https://github.com/scikit-build/scikit-build-core/blob/5389674500faad0c68ebfdc1f848102b34290ca6/src/scikit_build_core/build.py#L49

agoose77 replied:
It feels like `build` (the PyPI package) could do more here to check these dependencies, even if it's not bold enough to try and pip install them in the current environment. Have you any feelings there? This is not the first time that I've run into this.

henryiii replied:
`build` does, you are not using `build`, you are using `pip`. For `pip` it's an extra flag because adding it by default is not backward compatible: `python -m pip install --no-build-isolation --check-build-dependencies ./awkward-cpp`

henryiii replied:
(These are pretty light weight dependencies so I might add them by default at some point, especially for things like this. Though once scikit-build uses scikit-build-core it will likely include everything needed)

agoose77 replied:
Oh, you're right. I normally use `build`, so I assumed that was what I'd added into `CONTRIBUTING.md`. (I know why I put `pip` there, though; it's fewer tools to explain).

We can just update our `CONTRIBUTING.md` to make it clear what's needed; running without isolation is non-ideal for most cases anyway.

henryiii replied:
Just released the next version: https://github.com/scikit-build/scikit-build-core/releases/tag/v0.1.0a1 on PyPI.

You can do `scikit-build-core[pyproject]` to get the sdist/wheel dependencies directly. Might be a good idea for the requires, too.

ianna commented on src/awkward/_connect/cling.py:
```diff
@@ -4,6 +4,8 @@
 import json
 import re
 
+from awkward_cpp import libawkward
```
I like it :-)

ianna reviewed (approved):
Nice work! All tests pass. cpp tests do not link any unneeded libraries. There is one error, but I'm sure it's probably me. Thanks!

henryiii commented on CONTRIBUTING.md:
````diff
@@ -52,31 +52,66 @@ It is unnecessary to manually edit (rebase) your commit history. If, however, yo
+
+If you are working on the C++ components of Awkward Array, it might be more convenient to skip the build isolation step, which involves creating an isolated build environment. First, you must install the build requirements:
+```bash
+python -m pip install scikit-build-core pybind11 ninja cmake
````
```suggestion
python -m pip install "scikit-build-core[pyproject,color]" pybind11 ninja cmake
```

henryiii commented on CONTRIBUTING.md:
````diff
@@ -52,31 +52,66 @@ It is unnecessary to manually edit (rebase) your commit history. If, however, yo
+
+Then the installation can be performed without build isolation:
+```bash
+python -m pip install --no-build-isolation ./awkward-cpp
````
```suggestion
python -m pip install --no-build-isolation --check-build-dependencies ./awkward-cpp
```
Up to you.

agoose77 replied:
Nice, I wasn't aware of this `pip` flag.

henryiii commented on awkward-cpp/pyproject.toml:
```diff
@@ -0,0 +1,63 @@
+[build-system]
+requires = [
+	"scikit-build-core @ git+https://github.com/agoose77/scikit-build-core@27a60e553733ecffa7b411fe2760152555eaeefa",
```
```suggestion
    "scikit-build-core[pyproject]",
```

The pyproject is optional, but don't think there's a reason not to have it, and it could speed up the build slightly.

agoose77 committed:
docs: update CONTRIBUTING.md

Co-authored-by: Henry Schreiner <henry.fredrick.schreiner@cern.ch>

agoose77 committed:
chore: update awkward-cpp/pyproject.toml

Co-authored-by: Henry Schreiner <henry.fredrick.schreiner@cern.ch>

agoose77 committed:
docs: update CONTRIBUTING.md

Co-authored-by: Henry Schreiner <henry.fredrick.schreiner@cern.ch>

agoose77 committed:
docs: hide advanced information by default in CONTRIBUTING

pre-commit-ci[bot] committed:
style: pre-commit fixes

agoose77 committed:
Merge branch 'main' into agoose77/chore/split-awkward-package

jpivarski commented on CONTRIBUTING.md:
````diff
@@ -52,31 +52,73 @@ It is unnecessary to manually edit (rebase) your commit history. If, however, yo
+
+The C++ components can be installed by building the `awkward-cpp` package:
+```bash
+python -m pip install ./awkward-cpp
````
I think we always want to see the compilation output. Without `-vv`, a compiler or linker error would be hidden; pip would just say that it couldn't install and that isn't helpful. Anybody who is doing this step at least has the potential of having the compiler/linker fail, since the fully tested code-platform combinations are already wheels.

```suggestion
python -m pip install -vv ./awkward-cpp
```

henryiii replied:
I'd recommend one `-v`, but without it, pip _does_ dump the output if the build fails, it just hides it if it passes, so it seems to just "freeze" for a bit. "build" doesn't hide the output by default (and is both prettier and more succinct).

IIRC, some number of v's causes it to print insanely long "findlinks" output that has nothing to do with the build. So pretty sure you always only want one `-v`. The backend does not get any options to increase output with more v's, only pip (which a user doesn't normally care about).

jpivarski replied:
If it dumps the output on failure, then that's good enough now.

"Now" because the compilation is quick enough: a minute or so is a reasonable length of time for a no-information pause; five minutes wasn't. I guess I used to be antsy in that period of time, wanting to see the compilation percentage increase. I don't remember why I thought one `v` wasn't good enough. I must have settled on something that works.

So I'm going to resolve this conversation—don't add the "`-v`"; no change requested.

jpivarski reviewed (approved):
I followed the new

```bash
nox -s prepare
python -m pip install ./awkward-cpp
python -m pip install -e .
python -m pytest tests -n auto
```

procedure and had no trouble with the installation, except for the RDataFrame `DID NOT RAISE <class 'SystemError'>` bug on the _first_ run of the tests, but not any subsequent runs. It needs to be investigated, but not in this PR.

agoose77 committed:
ci: don't check versions for awkward-cpp

agoose77 committed:
ci: check awkward-cpp validity in awkward release

We can release awkward-cpp without any constraints; if the version is already on PyPI, PyPI will complain when we upload

agoose77 committed:
Merge remote-tracking branch 'origin/agoose77/chore/split-awkward-package' into agoose77/chore/split-awkward-package

agoose77 committed:
chore: move awkward version to pyproject.toml

agoose77 committed:
fix: fix nox test targets

agoose77 committed:
Merge branch 'main' into agoose77/chore/split-awkward-package

agoose77 committed:
ci: create headers archive

agoose77 committed:
Merge remote-tracking branch 'origin/agoose77/chore/split-awkward-package' into agoose77/chore/split-awkward-package

agoose77 committed:
chore: load project name from toml

pre-commit-ci[bot] committed:
style: pre-commit fixes

agoose77 committed:
chore: fix dev script

agoose77 committed:
ci: add note about why we set SOURCE_DATE_EPOCH

agoose77 committed:
chore: remove clean scripts in favor of nox

agoose77 committed:
docs: add notes about `nox -R`

agoose77 committed:
chore: fix auxiliary sdist includes

agoose77 committed:
chore: add `--endpoint` argument to `check-awkward-cpp-sdist-matches-pypi.py`

This can be used for testing

agoose77 committed:
chore: passthrough args in noxfile

agoose77 committed:
chore: include README.md in sdist

agoose77 committed:
chore: ensure sdists can be built into usable wheels

agoose77 committed:
chore: move kernel- files into git root

agoose77 committed:
chore: move header-only to git root

agoose77 committed:
perf: improve test generation speed

agoose77 committed:
chore: cleanup gitignore

agoose77 committed:
fix: re-order test generation

agoose77 committed:
fix: only use globals

agoose77 committed:
docs: add note about building wheels

agoose77 committed:
fix: package moved files appropriately

agoose77 committed:
chore: add "src/awkward/_connect/header-only" to clean

agoose77 committed:
ci: copy headers for docs building

agoose77 committed:
docs: add note about `nox -s prepare` for docs.

agoose77 committed:
ci: only lint PR title on changes to PR body

agoose77 committed:
Merge branch 'main' into agoose77/chore/split-awkward-package

agoose77 committed:
chore: set `tool.scikit-build.minimum-version`

jpivarski commented on awkward-cpp/pyproject.toml:
```diff
@@ -56,6 +56,9 @@ Chat = "https://gitter.im/Scikit-HEP/awkward-array"
 Releases = "https://github.com/scikit-hep/awkward-1.0/releases"
 
 
+[tool.scikit-build]
+minimum-version = "0.0"
```
Thanks, @agoose77!

henryiii replied:
I'd also recommend `sdist.reproducible = true`, just to clarify you expect that. If the default changes, the `minimum-version` will protect you, but might as well be explicit that that's critical to your use.

agoose77 committed:
Revert "ci: only lint PR title on changes to PR body"

This reverts commit e33c5edd11c5424c5d01a2fac25885cb0717a2c7.

henryiii commented on awkward-cpp/pyproject.toml:
```diff
@@ -0,0 +1,70 @@
+minimum-version = "0.0"
+
+[tool.scikit-build.sdist]
+include = [
```
This can be `sdist.include` to save making a new section.

agoose77 replied:
Cool, I hadn't realised that TOML had inline tables.

henryiii commented on awkward-cpp/pyproject.toml:
```diff
@@ -0,0 +1,70 @@
+    "Topic :: Software Development",
+    "Topic :: Utilities",
+]
+license = {text = "BSD-3-Clause"}
```
FYI, I think it's fine to just use the classifier.

agoose77 replied:
If this refers to the `license` field, I think I'd prefer to keep that explicitly in the metadata. I don't mind the duplication, whilst I do think we benefit from having it in a specified field.

agoose77 commented on awkward-cpp/pyproject.toml:
```diff
@@ -0,0 +1,70 @@
+    "tests-spec",
+    "tests-spec-explicit",
+    "tests-cpu-kernels"
+]
```
```suggestion
minimum-version = "0.0"
sdist.reproducible = true
sdist.include = [
    "header-only",
    "include/awkward/kernels.h",
    "src/awkward_cpp/_kernel_signatures.py",
    "tests-spec",
    "tests-spec-explicit",
    "tests-cpu-kernels"
]
```

agoose77 committed:
chore: update awkward-cpp/pyproject.toml
